packages/images/src/service.js

```diff
@@ -0,0 +1,199 @@
+        this.data = {};
+      } else {
+        throw err;
+      }
+    }
+  }
+  storeImage(id, meta) {
+    this.data[id] = meta;
+    fs.outputJSONSync(this.jsonPath, this.data);
+  }
+  getImage(id) {
+    return this.data[id];
+  }
+}
+
+const processResizeOptions = ({ width, height, fit = 'cover' }) => {
+  if (!width && !height) {
+    return null;
+  }
+  if (!VALID_FIT_TYPES.includes(fit)) {
+    return {
+      error: `Invalid fit option; can be one of ${VALID_FIT_TYPES.join(', ')}.`,
+    };
+  }
+  const widthInt = parseInt(width, 10);
+  const heightInt = parseInt(height, 10);
+
+  if (width && (isNaN(widthInt) || widthInt < 1)) {
+    return {
+      error: `Invalid width. Please provide a positive integer`,
+    };
+  }
+
+  if (height && (isNaN(heightInt) || heightInt < 1)) {
+    return {
+      error: `Invalid height. Please provide a positive integer`,
+    };
+  }
+
+  return { width: widthInt || undefined, height: heightInt || undefined, fit };
+};
+
+export class ImageService {
+  constructor({ port = 4001, protocol = 'http', host = 'localhost', path = './images' }) {
+    this.path = nodePath.resolve(path);
+    this.protocol = protocol;
+    this.port = port;
+    this.host = host;
+    this.database = new DataStore({ path: this.path });
+    this.sourcePath = nodePath.join(this.path, 'source');
+    this.transformsPath = nodePath.join(this.path, 'transforms');
+    fs.ensureDirSync(this.sourcePath);
+    fs.ensureDirSync(this.transformsPath);
+
+    const upload = multer({ dest: this.sourcePath });
+
+    this.app = express();
+
+    this.app.post('/upload', upload.single('image'), async (req, res) => {
+      const { originalname, path } = req.file;
+
+      res.status(201).json(this.storeImageInDatabase({ path, originalname }));
+    });
+
+    this.app.get('/image/:filename', async (req, res) => {
+      const { filename } = req.params;
+      const [id, format] = filename.split('.');
+
+      if (!id || !format) {
+        return res.status(404).send('Not Found (Invalid Filename)');
+      }
+
+      const data = this.database.getImage(id);
+
+      if (!data) {
+        return res.status(404).send('Not Found (No Matching Image)');
+      }
+
+      const originalPath = nodePath.join(this.sourcePath, `${id}.${data.format}`);
+      const resizeOptions = processResizeOptions(req.query);
+
+      if (!resizeOptions && format === data.format) {
+        return res.sendFile(originalPath);
+      }
+
+      let suffix = '';
+
+      if (resizeOptions) {
+        if (resizeOptions.error) {
+          return res.status(400).send(resizeOptions.error);
+        }
+
+        const md5sum = crypto.createHash('md5');
+        md5sum.update(JSON.stringify(resizeOptions));
+        const hash = md5sum.digest('hex');
+
+        suffix = `-${hash}`;
+      }
+
+      let transformedFilename = nodePath.join(this.transformsPath, `${id}${suffix}.${format}`);
+
+      if (!(await fs.exists(transformedFilename))) {
```

Comment:
IMO we should inverse this;  the simpler condition can terminate earlier.

Reply:
Not sure what you mean by this?